applications/FluidDynamicsApplication/custom_strategies/strategies/residualbased_simple_steady_scheme.h

```diff
@@ -267,32 +249,37 @@ class ResidualBasedSimpleSteadyScheme : public Scheme<TSparseSpace, TDenseSpace>
       if (rModelPart.GetCommunicator().MyPID() == 0)
         std::cout << "Computing OSS projections" << std::endl;
 
-      for (typename ModelPart::NodesContainerType::iterator itNode = rModelPart.NodesBegin();
-           itNode != rModelPart.NodesEnd(); itNode++)
-      {
-        noalias(itNode->FastGetSolutionStepValue(ADVPROJ)) = ZeroVector(3);
-        itNode->FastGetSolutionStepValue(DIVPROJ) = 0.0;
-        itNode->FastGetSolutionStepValue(NODAL_AREA) = 0.0;
+      const int number_of_nodes = rModelPart.NumberOfNodes();
+
+      #pragma omp parallel for
+      for (int i = 0; i < number_of_nodes; i++) {
+        ModelPart::NodeIterator it_node = rModelPart.NodesBegin() + i;
+        noalias(it_node->FastGetSolutionStepValue(ADVPROJ)) = array_1d<double,3>(3,0.0);
+        it_node->FastGetSolutionStepValue(DIVPROJ) = 0.0;
+        it_node->FastGetSolutionStepValue(NODAL_AREA) = 0.0;
       }
 
+      const int number_of_elements = rModelPart.NumberOfElements();
       array_1d<double, 3 > output;
-      for (typename ModelPart::ElementsContainerType::iterator itElem = rModelPart.ElementsBegin();
-           itElem != rModelPart.ElementsEnd(); itElem++)
-      {
-        itElem->Calculate(ADVPROJ, output, CurrentProcessInfo);
+
+      #pragma omp parallel for private(output)
+      for (int i = 0; i < number_of_elements; i++) {
+        ModelPart::ElementIterator it_elem = rModelPart.ElementsBegin() + i;
+        it_elem->Calculate(ADVPROJ,output,CurrentProcessInfo);
       }
+
       rModelPart.GetCommunicator().AssembleCurrentData(NODAL_AREA);
       rModelPart.GetCommunicator().AssembleCurrentData(DIVPROJ);
       rModelPart.GetCommunicator().AssembleCurrentData(ADVPROJ);
 
-      for (typename ModelPart::NodesContainerType::iterator itNode = rModelPart.NodesBegin();
-           itNode != rModelPart.NodesEnd(); itNode++)
-      {
-        if (itNode->FastGetSolutionStepValue(NODAL_AREA) == 0.0)
-          itNode->FastGetSolutionStepValue(NODAL_AREA) = 1.0;
-        const double Area = itNode->FastGetSolutionStepValue(NODAL_AREA);
-        itNode->FastGetSolutionStepValue(ADVPROJ) /= Area;
-        itNode->FastGetSolutionStepValue(DIVPROJ) /= Area;
+      #pragma omp parallel for
+      for (int i = 0; i < number_of_nodes; i++) {
+        ModelPart::NodeIterator it_node = rModelPart.NodesBegin() + i;
+        if (it_node->FastGetSolutionStepValue(NODAL_AREA) == 0.0)
```

Comment:
this seems dangerous ...

Reply:
This check used to be relevant when the mesher generated isolated nodes (not attached to any element). The solver will normally just ignore them but, without this check, here you would be dividing by zero and getting a nice NaN in your problem. Maybe a better soluition would be to deactivate such nodes, but this is the traditional kratos cure.

Reply:
I meant the double comparison against 0.0, I don't know the details of the algorithm

Reply:
In that case I think it is OK. The check is not to detect very small values (those should be OK because the variables you divide this by will also be very small), but for the case where there is no neighbouring element to contribute to the nodal area (in which case the nodal area should be exactly 0)